### PR TITLE
Feat/current filters view model

### PR DESCRIPTION
**Pull request by PLNech**
ViewModel + Tests.

### CODE REVIEW

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.algolia.instantsearch.core.event.EventViewModel
+import com.algolia.instantsearch.core.event.EventViewModelImpl
+import com.algolia.instantsearch.core.map.MapViewModel
+import com.algolia.search.model.filter.Filter
+
+
+class CurrentFiltersViewModel(
+    items: Map<String, Filter> = mapOf(),
+    val presenter: (Map<String, Filter>) -> Map<String, Filter> = { it }
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: That's the wrong place for a presenter. Presenter should only be passed in `connectView` methods

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+) : MapViewModel<String, Filter>(items),
+    EventViewModel<String> by EventViewModelImpl() {
+
+    fun getFilters(): Set<Filter> = item.values.toSet()
+
+    fun clearFilter(filter: Filter) {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: This method is only used in test and might be confusing for users. I think we can do without

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        clearFilter(item.entries.firstOrNull { it.value == filter }?.key)
+    }
+
+    fun clearFilter(identifier: String?) {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: how about a simpler approach:

```kotlin
public open class MapViewModel<K, V>(
    items: Map<K, V>
) : ItemViewModel<Map<K, V>>(items) {

    val onMapComputed: MutableList<(Map<K, V>) -> Unit> = mutableListOf()

    fun remove(key: K) {
        val map = item.toMutableMap().apply { remove(key) }

        onMapComputed.forEach { it(map) }
    }
}
```

Rendering the `CurrentFiltersViewModel` useful as a typealias only.

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import com.algolia.instantsearch.core.event.EventViewModelImpl
+import com.algolia.instantsearch.core.map.MapViewModel
+import com.algolia.search.model.filter.Filter
+
+
+class CurrentFiltersViewModel(
+    items: Map<String, Filter> = mapOf(),
+    val presenter: (Map<String, Filter>) -> Map<String, Filter> = { it }
+) : MapViewModel<String, Filter>(items),
+    EventViewModel<String> by EventViewModelImpl() {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: Reading what I wrote below, we won't need this

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    fun getFilters(): Set<Filter> = item.values.toSet()
+
+    fun clearFilter(filter: Filter) {
+        clearFilter(item.entries.firstOrNull { it.value == filter }?.key)
+    }
+
+    fun clearFilter(identifier: String?) {
+        identifier?.let {
+            item = presenter(item.toMutableMap().apply { remove(identifier) })
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: I think we should never let a `ViewModel` assign it's own member variable such as `item`. See the other `ViewModel`: Only `NumberViewModel` does it, but them the `item` `set` becomes `private` in favor of public function `applyBounds`

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt (hunk 3 above)
Comment (PLNech) [reply helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: :100:

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/map/MapView.kt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.algolia.instantsearch.core.map
+
+import com.algolia.instantsearch.core.event.EventView
+
+
+public interface MapView<K, V>: EventView<K> {
+
+    public fun setItems(items: Map<K, V>)
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/map/MapView.kt]: If `MapView<K, V>` extends `ItemView<Map<K, V>>`, we don't need `setItem` here

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/map/MapViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    val onMapComputed: MutableList<(Map<K, V>) -> Unit> = mutableListOf()
+
+    fun remove(key: K) {
+        val map = item.toMutableMap().apply { remove(key) }
+
+        onMapComputed.forEach { it(map) }
+    }
+
+    fun getValues() = item.values.toSet()
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/map/MapViewModel.kt]: Let's do the same thing than we want to do later with the alias:

`public val filters: Set<Filter> get() = item.values.toSet()`

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/selectable/segment/SelectableSegmentViewModel.kt [hunk 8]
```diff
@@ -1,12 +1,12 @@
...
 
-import com.algolia.instantsearch.core.item.ItemViewModel
+import com.algolia.instantsearch.core.map.MapViewModel
 import kotlin.properties.Delegates
 
 
 public open class SelectableSegmentViewModel<K, V>(
     items: Map<K, V>
-) : ItemViewModel<Map<K, V>>(items) {
+) : MapViewModel<K, V>(items) {
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/selectable/segment/SelectableSegmentViewModel.kt]: I don't think that's why now. That would give the `SelectableSegmentViewModel` the capability to do `remove`, and we don't want that yet.

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersConnectView.kt [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.algolia.instantsearch.helper.filter.current
+
+import com.algolia.search.model.filter.Filter
+
+
+public fun CurrentFiltersViewModel.connectView(view: CurrentFiltersView) {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersConnectView.kt]: Once rebased on `develop` we can refactor this function, but no probs 👍

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.algolia.instantsearch.helper.filter.current
+
+import com.algolia.instantsearch.core.map.MapViewModel
+import com.algolia.search.model.filter.Filter
+
+
+typealias CurrentFiltersViewModel = MapViewModel<String, Filter>
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/current/CurrentFiltersViewModel.kt]: Missing `public`

File: helper/src/commonTest/kotlin/filter/current/TestCurrentFiltersViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        viewModel.remove(red)
+        viewModel.item shouldEqual mapOf(green to filterGreen)
+    }
+
+    @Test
+    fun clearFilterTriggersListeners() {
+        val viewModel = CurrentFiltersViewModel(mapOf(green to filterGreen))
+        var triggered = false
+        viewModel.onMapComputed += {
```
Comment (q-litzler) [inline helper/src/commonTest/kotlin/filter/current/TestCurrentFiltersViewModel.kt]: Missing line

File: helper/src/commonTest/kotlin/filter/current/TestCurrentFiltersViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            triggered = true
+        }
+
```
Comment (q-litzler) [inline helper/src/commonTest/kotlin/filter/current/TestCurrentFiltersViewModel.kt]: Extra line

**Comment by q-litzler** [comment]
Note: We should rather test `MapViewModel` than `CurrentFiltersViewModel`

**Pull request merged**